php/libraries/Candidate.class.inc

```diff
@@ -699,52 +699,92 @@ class Candidate
+            }
+        } else {
+            do {
+                $generator->nextRand();
+                $result = $generator->cur();
+            } while (!is_null($result) && $is_colliding_callback($result));
+        }
+
+        if (is_null($result)) {
+            $attempts = $generator->getAttempts();
+            $seqCount = $generator->sequenceCount();
+            $error    = "Could not generate {$name} after {$attempts} attempts. "
+                . "All {$seqCount} possible {$name} have been generated. "
+                . "Please set {$name} generator definition "
+                . "to allow more {$name} to be generated. ";
+            if (!$is_sequential) {
+                $error .= "(expected number of candidates)^3 is recommended "
+                    . "for random generation.";
+            }
+            throw new Exception($error);
+        }
+        return $result;
+    }
+    /**
     * Generates a new PSCID in the format defined by the config file
     *
-    * @param string $prefix The prefix to use for the PSCID
+    * @param array $static_dict Example: array("siteAbbrev"=>"DCC")
     *
-    * @return string Project candidate id in format defined by config.xml
+    * @return string id in format defined by config.xml
     */
-    static function _generatePSCID($prefix)
+    static function _generatePSCID($static_dict)
     {
-        // get a database connection
         $factory = NDB_Factory::singleton();
         $db      = $factory->database();
+        $config  = $factory->config();
 
-        $config = $factory->config();
-
-        $PSCIDSettings = $config->getSetting('PSCID');
-        $pscidArrays   = Candidate::generateArrays($PSCIDSettings, $prefix);
-
-        // run one first time, then loop until the pscid is not used
-        do {
-            // pick a new pscid
-            if ($PSCIDSettings['generation'] == 'sequential') {
-                if (!empty($pscid)) {
-                    $pscid++;
-                } else {
-                    $pscid = Candidate::generateFirstStringFromArrays($pscidArrays);
-                }
-            } else {
-                $pscid = Candidate::generateRandomStringFromArrays($pscidArrays);
-            }
-
-            if ($pscid == '') {
-                throw new LorisException(
-                    "Well this is embarrassing, our automated PSCID generator "
-                    . "is about to enter an infinite loop."
+        return self::_generateId(
+            "PSC Id",
+            $config->getSetting("PSCID"),
+            $static_dict,
+            function ($needle) use ($db) {
+                $colliding = $db->pselectOne(
+                    "
+                        SELECT EXISTS (
+                            SELECT
+                                *
+                            FROM
+                                candidate
+                            WHERE
+                                PSCID = :needle
```

Comment:
`SELECT EXISTS (SELECT * ...)` is a valid query and it captures the intention better than the proposed query.